#! /bin/sh /usr/share/dpatch/dpatch-run
## fix buffer overflow in recursor 
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix exit status

@DPATCH@
--- pdns-2.9.17.orig/pdns/pdns_recursor.cc
+++ pdns-2.9.17/pdns/pdns_recursor.cc
@@ -563,7 +563,7 @@
 	    if(bytes==1)
 	      i->state=TCPConnection::BYTE1;
 	    if(bytes==2) { 
-	      i->qlen=(i->data[0]<<8)+i->data[1];
+	      i->qlen=(((unsigned char)i->data[0]) << 8)+ (unsigned char)i->data[1];
 	      i->bytesread=0;
 	      i->state=TCPConnection::GETQUESTION;
 	    }
@@ -577,7 +577,7 @@
 	    int bytes=read(i->fd,i->data+1,1);
 	    if(bytes==1) {
 	      i->state=TCPConnection::GETQUESTION;
-	      i->qlen=(i->data[0]<<8)+i->data[1];
+	      i->qlen=(((unsigned char)i->data[0]) << 8)+ (unsigned char)i->data[1];
 	      i->bytesread=0;
 	    }
 	    if(!bytes || bytes < 0) {

